modules/core/arrow-annotations/src/main/java/arrow/product.kt

```diff
@@ -0,0 +1,8 @@
+package arrow
+
+import kotlin.annotation.AnnotationRetention.SOURCE
+import kotlin.annotation.AnnotationTarget.CLASS
+
+@Retention(SOURCE)
+@Target(CLASS)
+annotation class product
```

Comment:
I guess product refers to product types but the annotation should probably reflect what's it really doing for the end user. Why not calling it something like `@derive` or something similar?

Reply:
@product is ment to do a lot more than derivation. For example the `tupled` methods are not directly related to derivation. The idea is to have under generic `@product`, `@coproduct` and that add the features inside those annotations opt out. `@product(deriving = Show.class, isoExtensions = HList)` This would tell to just generate `Show` and the `HList` extension for advanced use cases.
At the moment granular support for a la carte derivation is not built in since this is just a first approach.